### PR TITLE
Fix non-local variables in rustup-init

### PR DESCRIPTION
**Pull request by tesuji**
Hope this fix those issues ([1], [2]) pointed by @ehuss in #1582 .

[1]: https://github.com/rust-lang/rustup.rs/pull/1582#discussion_r245556881
[2]: https://github.com/rust-lang/rustup.rs/pull/1582#discussion_r245556864

### CODE REVIEW

**Comment by tesuji** [comment]
Rebased on master.

**Comment by nrc** [comment]
@ehuss would you mind taking a look please?

File: rustup-init.sh [hunk 1]
```diff
@@ -175,9 +172,10 @@ get_endianness() {
 }
 
 get_architecture() {
-    local _ostype _cputype
+    local _ostype _cputype _bitness _arch
     _ostype="$(uname -s)"
     _cputype="$(uname -m)"
+    _bitness="$(get_bitness)"
```
Comment (ehuss) [inline rustup-init.sh]: This now runs on all platforms, which will fail on some with the following error.
```
head: /proc/self/exe: No such file or directory
rustup: unknown platform bitness
```
I suspect it was only run on linux for that reason.

**Review by ehuss — CHANGES_REQUESTED** [review]

**Comment by ehuss** [comment]
At this time I don't think we want to make this many changes to the init script.  If you can narrow it down to just adding the appropriate `local` declarations, I think that would be good.  Thanks for taking a look, though!

File: rustup-init.sh [hunk 2]
```diff
@@ -15,14 +15,13 @@
 
 set -u
 
-if [ -z "$RUSTUP_UPDATE_ROOT" ]; then
-    RUSTUP_UPDATE_ROOT="https://static.rust-lang.org/rustup"
-fi
+RUSTUP_UPDATE_ROOT=${RUSTUP_UPDATE_ROOT:-"https://static.rust-lang.org/rustup"}
```
Comment (lzutao) [inline rustup-init.sh]: This change fixes Undefined Variables error in `RUSTUP_UPDATE_ROOT` because of setting `set -u`.

File: rustup-init.sh [hunk 3]
```diff
@@ -268,23 +266,15 @@ get_architecture() {
             ;;
 
         mips)
-            _cputype="$(get_endianness $_cputype "" 'el')"
+            _cputype=$(get_endianness mips '' el)
             ;;
```
Comment (lzutao) [inline rustup-init.sh]: I think that change is easier to read.

File: rustup-init.sh [hunk 4]
```diff
@@ -268,23 +266,15 @@ get_architecture() {
...
-            _bitness="$(get_bitness)"
-            if [ "$_bitness" -eq 32 ]; then
-                if [ "$_ostype" = "unknown-linux-gnu" ]; then
-                    # 64-bit kernel with 32-bit userland
-                    # endianness suffix is appended later
-                    _cputype=mips
-                fi
-            else
+            if [ "$_bitness" -eq 64 ]; then
                 # only n64 ABI is supported for now
```
Comment (lzutao) [inline rustup-init.sh]: Does mips64 have 16-bits?

File: rustup-init.sh [hunk 5]
```diff
@@ -15,14 +15,13 @@
 
 set -u
 
-if [ -z "$RUSTUP_UPDATE_ROOT" ]; then
-    RUSTUP_UPDATE_ROOT="https://static.rust-lang.org/rustup"
-fi
+RUSTUP_UPDATE_ROOT=${RUSTUP_UPDATE_ROOT:-"https://static.rust-lang.org/rustup"}
 
 #XXX: If you change anything here, please make the same changes in setup_mode.rs
+version='rustup-init 1.16.0 (beab5ac2b 2018-12-06)'
```
Comment (dwijnand) [inline rustup-init.sh]: Should we keep this local to `usage`?

**Comment by nrc** [comment]
ping @lzutao do you have any update here? Could you reply to the questions inline please?

File: rustup-init.sh (hunk 5 above)
Comment (tesuji) [reply rustup-init.sh]: Maybe rename it to `rustup_version`?

Comment (tesuji) [reply rustup-init.sh]: The reason I split it to `version` is to easy to spot this need to be updated. People seems to forget to update this field.

**Comment by tesuji** [comment]
@nrc I think this PR make a small change to init script. Should I split this change to another PR?
Cause one important point of this change is to fix **Undefined Variables** error 
in `RUSTUP_UPDATE_ROOT` because of setting `set -u`.

File: rustup-init.sh (hunk 5 above)
Comment (dwijnand) [reply rustup-init.sh]: Ah, good reason.  I'm fine with it staying as "version".

**Comment by kinnison** [comment]
My rough feeling is that we can probably safely merge this, though I would like to be reassured that it behaves sensibly on MIPS since that seems to be the crux of the change.  If you're not able to provide that reassurance then yes, making this PR only about fixing the actual bug you describe and then a separate PR can be about solving the MIPS related stuff.

**Comment by tesuji** [comment]
Force push to only include local stuff fix.

Thanks @kinnison, I would open another PR for other stuffs.

**Review by kinnison — APPROVED** [review]
I think this now LGTM. Do you agree @dwijnand ? @nrc ?

**Review by dwijnand — APPROVED** [review]
yes

**Pull request merged**